rust/src/measurements/make_private_lazyframe/group_by/mod.rs

```diff
@@ -21,136 +33,184 @@ use polars_plan::prelude::GroupbyOptions;
-    }
-
-    if apply.is_some() {
-        return fallible!(MakeMeasurement, "Apply is not supported in logical plan");
-    }
-
-    let keys = keys
-        .iter()
-        .map(|e| {
-            Ok(match e {
-                Expr::Column(name) => vec![(*name).to_string()],
-                Expr::Columns(names) => names.iter().map(|s| s.to_string()).collect(),
-                e => {
-                    return fallible!(
-                        MakeMeasurement,
-                        "Expected column expression in keys, found {:?}",
-                        e
-                    )
-                }
-            })
-        })
-        .collect::<Fallible<Vec<_>>>()?
-        .into_iter()
-        .flatten()
-        .collect::<BTreeSet<_>>();
+    let grouping_columns = match_grouping_columns(keys.clone())?;
 
     let margin = middle_domain
         .margins
-        .get(&keys)
-        .ok_or_else(|| err!(MakeMeasurement, "Failed to find margin for {:?}", keys))?
-        .clone();
+        .get(&grouping_columns)
+        .cloned()
+        .unwrap_or_default();
 
     let expr_domain = ExprDomain::new(
         middle_domain.clone(),
         ExprContext::Aggregate {
-            grouping_columns: keys,
+            grouping_columns: grouping_columns.clone(),
         },
     );
 
-    let t_group_by = Transformation::new(
-        expr_domain.clone(),
-        expr_domain.clone(),
-        Function::new(Clone::clone),
-        middle_metric.clone(),
-        PartitionDistance(middle_metric.clone()),
-        StabilityMap::new_fallible(move |&d_in| {
+    let m_exprs = make_basic_composition(
+        aggs.into_iter()
+            .map(|expr| {
+                make_private_expr(
+                    expr_domain.clone(),
+                    PartitionDistance(middle_metric.clone()),
+                    output_measure.clone(),
+                    expr.clone(),
+                    global_scale,
+                )
+            })
+            .collect::<Fallible<_>>()?,
+    )?;
+
+    let f_comp = m_exprs.function.clone();
+    let privacy_map = m_exprs.privacy_map.clone();
+    let dp_exprs = m_exprs.invoke(&(input, all()))?;
+
+    if threshold.is_some() && expr_threshold.is_some() {
+        return fallible!(MakeMeasurement, "two thresholds set: one via the constructor and another via filtering. Please remove one");
+    }
+
+    let threshold = if margin.public_info.is_some() {
+        None
+    } else if let Some((name, threshold_value)) = expr_threshold {
+        let noise = find_len_expr(&dp_exprs, Some(name.as_str()))?.1;
+        Some((name, noise, threshold_value))
+    } else if let Some(threshold_value) = threshold {
+        let (name, noise) = find_len_expr(&dp_exprs, None)?;
+        Some((name, noise, threshold_value))
+    } else {
+        // TODO: link to documentation page showing how to filter.
```

Comment:
I like to include an issue URL on TODO notes, but if you've got this in your head, not critical.